src/librustc_privacy/lib.rs

```diff
@@ -1504,9 +1517,29 @@ impl<'a, 'tcx: 'a> SearchInterfaceForPrivateItemsVisitor<'a, 'tcx> {
                 self.tcx.lint_node(lint::builtin::PRIVATE_IN_PUBLIC, node_id, self.span,
                                    &format!("{} (error {})", msg, err_code));
             }
+
         }
+
         false
     }
+
+    /// An item is 'leaked' from a private dependency if all
+    /// of the following are true:
+    /// 1. It's contained within a public type
+    /// 2. It does not come from a crate marked as public
+    fn leaks_private_dep(&self, item_id: DefId) -> bool {
+        // Never do any leak checking if the feature is not enabled
+        if !self.tcx.features().public_private_dependencies {
+            return false
+        }
+        let ret = self.required_visibility == ty::Visibility::Public &&
+            !item_id.is_local() &&
```

Comment:
Nit: this condition is probably unnecessary.
(Also it could be moved to `public_crates` collection, to make sure that local crate *certainly* doesn't end up in `public_crates`.)